agent/engine/dockerclient/dockerclientfactory.go

```diff
@@ -110,7 +120,15 @@ func (f *factory) getClient(version DockerVersion) (dockeriface.Client, error) {
 // findDockerVersions loops over all known API versions and finds which ones
 // are supported by the docker daemon on the host
 func findDockerVersions(endpoint string) map[DockerVersion]dockeriface.Client {
-	clients := make(map[DockerVersion]dockeriface.Client)
+	// if the client version returns a MinAPIVersion and APIVersion, then use it to return
+	// all the Docker clients between MinAPIVersion and APIVersion
+	clients, err := findDockerVersionsfromMinMaxVersions(endpoint)
+	if err != nil {
+		log.Infof("Unable to get Docker clients from min API version: %v", err)
+	} else {
+		return clients
```

Comment:
The second concern is tangential. Putting this here puts a hard branch in our detection logic, and I think we should consider composing it differently.

```golang
// As written, the code boils down to this
clients, err := findDockerVersionsOneWay()
if err == nil {
    return clients
}
clients, err := findDockerVersionsADifferentWay()
```
I'd rather see us handle the detection in the same place using existing logic instead of branching like this. One suggestion would be to look at the `version_support_{platform}` files and the Min version variable. You could write a function that does something like this:

```golang
 // Use either the docker minimum API version or rely on the version set in version_support_{platform}.go
func getMinVersion() DockerVersion {
    if functionThatChecksIfDockerHasMinAPIParam() {
        return functionThatExtractsMinAPIFromDockerAPI()
    } else {
        return minDockerAPIVersion
    }
}

Reply:
Oh &mdash; You'd want to write a function to check for the max version, too :)

Reply:
What would be the default max version here be?